### PR TITLE
add support for incremental step stats calculations

### PR DESCRIPTION
**Pull request by prha**
## Summary & Motivation
This is the counterpart of https://github.com/dagster-io/dagster/pull/26550, but for step stats instead of run stats.

The step stats calculations are a little more complicated because it relies on intermediate state in between dagster events to be held.  This includes the `previous_attempt_start` time for a given step.  This PR introduces a field in the snapshot to carry over this state, which should only be used in the incremental calculation.

Another change made here is that marker events (e.g. resource init/teardown) can be associated with a step, but generally occur before the step has started.  That means that previously step snapshots did not include these marker events if the step had not yet started.  This PR changes that behavior so that every step will have a snapshot if it has a marker event, even if it has not started.   The output for incremental vs completed run snapshots should match exactly.

## How I Tested These Changes
BK

### CODE REVIEW

**Comment by prha** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/26577?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#26577** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/26577?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/26577?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#26550** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/26550?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `master`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: python_modules/dagster/dagster_tests/storage_tests/test_event_log.py [hunk 1]
```diff
@@ -326,3 +330,48 @@ def _ops():
...
+    assert op_failure_stats
+    assert op_failure_stats.attempts == 4
+    assert len(op_failure_stats.attempts_list) == 4
+
+    # build up run stats through incremental events
+    incremental_step_stats = None
+    for event in events:
+        incremental_step_stats = build_run_step_stats_from_events(
+            result.run_id, [event], incremental_step_stats
+        )
```
Comment (gibsondan) [inline python_modules/dagster/dagster_tests/storage_tests/test_event_log.py:375]: should we do any checkups of the partially constructed stats partway through to make sure they match what we expect? like when its picked up a marker start but not yet a marker end? or is the end part where it verifies the final results are correct sufficient?

**Review by gibsondan — APPROVED** [review]

File: python_modules/dagster/dagster_tests/storage_tests/test_event_log.py (hunk 1 above)
Comment (prha) [reply python_modules/dagster/dagster_tests/storage_tests/test_event_log.py:375]: I thought about it, but I'm not sure it really matters?

Comment (prha) [reply python_modules/dagster/dagster_tests/storage_tests/test_event_log.py]: Okay, scratch that... one major consumer of step stats is the k8s executor which uses the presence of step stats as a proxy for started steps.  I updated the PR to revert back to the original behavior and created a helper function to power the partial snapshots.

Also converted the NT => record.

**Review by gibsondan — APPROVED** [review]

**Pull request merged**